### DISCUSSION TITLE
Enforcing nameof Usage with [MandatoryNameof] Attribute

### DISCUSSION THREAD

**Discussion by 2Xpro-pop** [dotnet/csharplang#8002] [Language Ideas]
I’d like to discuss a proposal for a new attribute called [MandatoryNameof], aimed at enhancing code safety and readability. This attribute ensures the use of nameof for method parameters, reducing the reliance on magic strings.

## The Proposal

Introduce [MandatoryNameof], an attribute to enforce compile-time checks for nameof usage:

```csharp
[AttributeUsage(AttributeTargets.Parameter)]
public class MandatoryNameofAttribute : Attribute
{
    public MandatoryNameofTarget Target { get; set; } = MandatoryNameofTarget.All;
    public Type? Type { get; set; } = null;

    public MandatoryNameofAttribute() { }
}
```

## How It Works

- Target: Specifies the type of code elements (e.g., Property, Method, Event, Field, or Type) that the nameof should reference. All is the default, allowing any type.
- Type: (Optional) When set, restricts the nameof reference to members of the specified type, enhancing type safety.

## Example Use Case
```csharp
public void RegisterEventHandler([MandatoryNameof(Target = MandatoryNameofTarget.Event, Type = typeof(MyEvents))] string eventName)
{
    // Event handler registration logic
}
```

This enforces that eventName must be the name of an event defined in the MyEvents class.

**Comment by CyrusNajmabadi** [comment]
This does not need any language change. It is already possible for you to create such an attribute today and then check it with an analyzer you write :-)

This is a core example of the type of scenario analyzers set out to provide a solution for :-)
